From efba720910e09b719beabcdec4d4126498e23f28 Mon Sep 17 00:00:00 2001
From: Zhuang Jin Can <jin.can.zhuang@intel.com>
Date: Thu, 5 Mar 2015 16:52:01 +0800
Subject: [PATCH] xhci: workaround for ssic PME issue

There's a limitation in SSIC that when the port is in Rx.detect, xHCI
can't enter D3 due to PME storm caused by the port. To workaround the
limitation, the port is required to set to unused before D3 entry and
set to used after D3 exit. This is the same as what has been done in
previous patch "xhci-pci: sw workaround for multiple HcRST hang and
spurious PME issue". The previous patch only handled one SSIC port
while this patch handles both ssic ports in cherrytrail.

Change-Id: Id78d2b6405ea7dd0290d961fc37c7238c3a5ea24
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-26932
Signed-off-by: Zhuang Jin Can <jin.can.zhuang@intel.com>
---
 drivers/usb/host/xhci-intel-cap.c | 37 ++++++++++++++++++++-----------------
 drivers/usb/host/xhci-intel-cap.h |  9 ++++++++-
 2 files changed, 28 insertions(+), 18 deletions(-)

diff --git a/drivers/usb/host/xhci-intel-cap.c b/drivers/usb/host/xhci-intel-cap.c
index 679e4d0..35fcf74 100644
--- a/drivers/usb/host/xhci-intel-cap.c
+++ b/drivers/usb/host/xhci-intel-cap.c
@@ -153,7 +153,7 @@ EXPORT_SYMBOL_GPL(xhci_intel_clr_internal_pme_flag);
    pme and HCRST hangs issue */
 void xhci_intel_ssic_port_unused(struct xhci_hcd *xhci, bool unused)
 {
-	int ext_offset;
+	int ext_offset, i;
 	void __iomem *ssic_port_cfg;
 	u32 data;
 
@@ -166,21 +166,24 @@ void xhci_intel_ssic_port_unused(struct xhci_hcd *xhci, bool unused)
 		return;
 
 	ssic_port_cfg = &xhci->cap_regs->hc_capbase +
-			((ext_offset + SSIC_PORT1_CFG2) >> 2);
-
-	data = readl(ssic_port_cfg);
-	data &= ~PROG_DONE;
-	writel(data, ssic_port_cfg);
-
-	data = readl(ssic_port_cfg);
-	if (unused)
-		data |= SSIC_PORT_UNUSED;
-	else
-		data &= ~SSIC_PORT_UNUSED;
-	writel(data, ssic_port_cfg);
-
-	data = readl(ssic_port_cfg);
-	data |= PROG_DONE;
-	writel(data, ssic_port_cfg);
+			((ext_offset + SSIC_PORT_CFG2) >> 2);
+	for (i = 0; i < SSIC_PORT_NUM; i++) {
+		data = readl(ssic_port_cfg);
+		data &= ~PROG_DONE;
+		writel(data, ssic_port_cfg);
+
+		data = readl(ssic_port_cfg);
+		if (unused)
+			data |= SSIC_PORT_UNUSED;
+		else
+			data &= ~SSIC_PORT_UNUSED;
+		writel(data, ssic_port_cfg);
+
+		data = readl(ssic_port_cfg);
+		data |= PROG_DONE;
+		writel(data, ssic_port_cfg);
+
+		ssic_port_cfg += SSIC_PORT_CFG2_OFFSET;
+	}
 }
 EXPORT_SYMBOL_GPL(xhci_intel_ssic_port_unused);
diff --git a/drivers/usb/host/xhci-intel-cap.h b/drivers/usb/host/xhci-intel-cap.h
index d3f2757..8b805eb 100644
--- a/drivers/usb/host/xhci-intel-cap.h
+++ b/drivers/usb/host/xhci-intel-cap.h
@@ -30,7 +30,14 @@
 #define DUAL_ROLE_CFG1		0x6c
 #define SW_MODE			(1 << 29)
 
-#define SSIC_PORT1_CFG2		0x3c
+#define SSIC_PORT_NUM		2
+
+/* SSIC Configuration Register 2
+ * Address Offset: 0Ch-0Fh
+ * Port 1 ... N : 0Ch, 3Ch, ... ,(0Ch + (NumSSICPorts-1)*30h)
+ */
+#define SSIC_PORT_CFG2		0xc
+#define SSIC_PORT_CFG2_OFFSET	0x30
 #define PROG_DONE		(1 << 30)
 #define SSIC_PORT_UNUSED	(1 << 31)
 
-- 
1.9.1

